File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.java
/*
TC: O(n2)
SC: O(1)
Approach: for every index, store the count of the longest increasing subsequence and increase
it along the way, keep a currMax to track it and when you are done, just return the map.get(currMax)
*/

class Solution {
    public int findNumberOfLIS(int[] nums) {
        int currMax = 1;
        Map<Integer,Integer>[] count = new HashMap[nums.length];
        Map<Integer,Integer> cum = new HashMap<>();
        int[] dp = new int[nums.length];
        dp[0] = 1;
        if(count[0] == null) {
            count[0] = new HashMap<>();
            count[0].put(1,1);
            cum.put(1,1);
        }
        for(int i = 1; i < nums.length; i++) {
            int maxLen = 1;
            if(count[i] == null) {
                count[i] = new HashMap<>();
                count[i].put(1,1);
                cum.put(1,cum.getOrDefault(1,0) + 1);
            }
            for(int j = 0; j < i; j++) {
                if(nums[j] < nums[i]) {
                    count[i].put(dp[j] + 1,count[i].getOrDefault(dp[j]+1,0) + count[j].get(dp[j]));
                    cum.put(dp[j] + 1,cum.getOrDefault(dp[j] + 1,0) + count[j].get(dp[j]));
                    maxLen = Math.max(maxLen,dp[j] + 1);
                }
            }
            dp[i] = maxLen;
            currMax = Math.max(maxLen,currMax);
        }
        return cum.get(currMax);
    }
}

File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.py
class Solution(object):
    def findNumberOfLIS(self, nums):
        currMax = 1
        count = [defaultdict(int) for _ in range(len(nums))]
        cum = defaultdict(int)
        dp = [1] * len(nums)
        
        count[0][1] = 1
        cum[1] = 1

        for i in range(1, len(nums)):
            maxLen = 1
            count[i][1] = 1
            cum[1] += 1
            
            for j in range(i):
                if nums[j] < nums[i]:
                    count[i][dp[j] + 1] += count[j][dp[j]]
                    cum[dp[j] + 1] += count[j][dp[j]]
                    maxLen = max(maxLen, dp[j] + 1)
                    
            dp[i] = maxLen
            currMax = max(currMax, maxLen)
        
        return cum[currMax]

            
        